### DISCUSSION TITLE
How to run vsgImGui example?

### DISCUSSION THREAD

**Discussion by symBroseph** [vsg-dev/VulkanSceneGraph#1116] [Q&A] [answered]
Hi,

I tried running the vsgImGui example after building everything with the vsgFramework (what was current as of 2024-2-28), but although everything compiled just fine, the vsgImGui example does not show any UI elements at all:

![Capture](https://github.com/vsg-dev/VulkanSceneGraph/assets/81342925/801df889-a794-4861-89c8-c394d83525ea)
*vsgimgui_example.exe run with teapot.vsgt*

Upon further inspection I realized that nothing in the `MyGui::record` and `MyGui::compile` functions is ever run. I don't know why this is, as I can see where the MyGui instance is added to the rendergraph. There are no errors either, just no UI. 

Why is this? The source code is just what's in the vsgExamples repo (plus some `cout` statements for basic debugging).

**Comment by robertosfield** [comment]
I just tried the vsgimgui_example on this Ubuntu 22.04 with VSG, vsgImGui and vsgExamples master and I get the same result - the 3d model but no ImGui.  It's late in the evening here so I'll look into it further tomorrow.  It'll be a regression due to recent changes to VSG or vsgImGui so I'll use git bisect to figure what has caused the regression.

**Comment by robertosfield** [comment]
@symBroseph I have tested various version of vsgImGui and found they all fail to show any graphics when built against VSG masterm but works with VulkanSceneGraph v1.1.0 or earlier.  I am now using git bisect on the VSG to see which commit since v1.1.0 caused the regression in vsgImGui.

**Comment by robertosfield** [comment]
The result of my git bisect usage is the commit that works with vsgImGui is: 311c893c1657832b1ba5a1046d6bfb4b2974250c, and 

And the the one that breaks it is: e159e66d9bb2946936ae2422c926f06633230235

@AnyOldName3 this commit is from your #1106 PR, any ideas why this might cause a regression in vsgImGui?

**Reply by robertosfield** [reply]
I have been investigating what the difference are between vsgImGui running before and after the problem commit and the good VSG version the vsgImGui::Texture::compile() is called as required, but with the bad commit this doesn't happen.

I don't yet know why the use of std::forward is causing the difference for vsgImGui.

**Comment by robertosfield** [comment, marked as answer]
I have tracked down the regression to vsgImGui::RenderImGui compatibility with the new changes to vsg::Inherit, the fix is now merged with vsgImGui master:

https://github.com/vsg-dev/vsgImGui/commit/7c05fb691370d3b6ac6e2837f23b802af717c7dd
